File: 22_Generate Parentheses.java
// Time Complexity: O(n * C(2n,n))
// Space Complexity: O(n)

class Solution {
    private List<String> res = new ArrayList<>();
    private StringBuilder sb;

    public List<String> generateParenthesis(int n) {
        this.sb = new StringBuilder(n*2);
        bt(n, 0);
        return res;
    }

    private void bt(int openCanBePlaced, int closeCanBePlaced) {
        // end case
        if (openCanBePlaced == 0 && closeCanBePlaced == 0) {
            res.add(sb.toString());
            return;
        }
        // process
        // backtrack
        if (openCanBePlaced > 0) {
            sb.append('(');
            bt(openCanBePlaced-1, closeCanBePlaced+1);
            sb.deleteCharAt(sb.length() - 1);
        }
        if (closeCanBePlaced > 0) {
            sb.append(')');
            bt(openCanBePlaced, closeCanBePlaced-1);
            sb.deleteCharAt(sb.length() - 1);
        }
    }
}

File: 22_Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def bt(openP, closeP):
            # end case
            if openP == n and closeP == n:
                res.append(''.join(curPath))
                return
            # process: choose ) or not
            if openP > closeP: 
                curPath.append(')')
                bt(openP, closeP + 1)
                curPath.pop()
            if openP < n:
                curPath.append('(')
                bt(openP + 1, closeP)
                curPath.pop()

        res, curPath = [], []
        bt(0, 0)
        return res